src/main/java/org/manifold/compiler/front/MappedArray.java

```diff
@@ -0,0 +1,163 @@
+package org.manifold.compiler.front;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+
+import java.util.*;
+
+public class MappedArray<K, V> implements Iterable<MappedArray<K, V>.Entry> {
+
+  private List<V> data;
+  private BiMap<K, Integer> keyedIndices = HashBiMap.create();
+
+  public MappedArray(Map<K, V> data) {
+    this.data = new ArrayList<>(data.size());
+
+    int i = 0;
+    for (Map.Entry<K, V> entry : data.entrySet()) {
+      this.data.add(entry.getValue());
+      keyedIndices.put(entry.getKey(), i);
+    }
+  }
+
+  public MappedArray(List<V> data) {
+    this.data = data;
```

Comment:
Why not? Wouldn't a shallow copy mean that you now have two `MappedArray`s sharing the same data, such that modifying one would inadvertently cause the other to change? (See also how `copyOf()` is implemented)